src/bottom-bar/use-on-complete-callback.js

```diff
@@ -0,0 +1,166 @@
+)
+
+function hasViolations(commentRequiredViolations, validationRuleViolations) {
+    return (
+        commentRequiredViolations.length ||
+        Object.keys(validationRuleViolations).length
+    )
+}
+
+function useOnCompleteWhenValidRequiredClick() {
+    const { mutateAsync: setFormCompletion } = useSetFormCompletionMutation()
+    const setRightHandPanel = useSetRightHandPanel()
+    const validate = useImperativeValidate()
+
+    return () =>
+        validate()
+            // Show this alert message and skip the next "then" block
+            .catch((e) => {
+                console.error(e)
+                throw new Error(validationFailedMessage)
+            })
+            // the validation request succeeded
+            .then(({ commentRequiredViolations, validationRuleViolations }) => {
+                // if the form is invalid, show the sidebar and show an
+                // alert to the user
+                if (
+                    hasViolations(
+                        commentRequiredViolations,
+                        validationRuleViolations
+                    )
+                ) {
+                    setRightHandPanel(validationResultsSidebarId)
+                    return Promise.reject(
+                        new Error(
+                            i18n.t("The form can't be completed while invalid")
+                        )
+                    )
+                }
+                // otherwise complete the form
+                else {
+                    return setFormCompletion({ completed: true }).catch((e) => {
+                        console.error(e)
+                        throw new Error(completingFormFailedMessage)
+                    })
+                }
+            })
+}
+
+function useOnCompleteWhenValidNotRequiredClick() {
+    const { mutateAsync: setFormCompletion } = useSetFormCompletionMutation()
+    const setRightHandPanel = useSetRightHandPanel()
+    const validate = useImperativeValidate()
+    const { show: showWarningAlert } = useAlert((message) => message, {
+        warning: true,
+    })
+
+    return () => {
+        return Promise.all([
+            validate()
+                .then(
+                    ({
+                        commentRequiredViolations,
+                        validationRuleViolations,
+                    }) => {
+                        if (
+                            hasViolations(
+                                commentRequiredViolations,
+                                validationRuleViolations
+                            )
+                        ) {
+                            setRightHandPanel(validationResultsSidebarId)
+                        }
+                    }
+                )
+                .catch((e) => {
+                    console.error(e)
+                    // For now this can fail "silently".
+                    // ES6 Promises don't have a mechanism for reporting
```

Comment:
```js
Promise.allSettled() 
```
Would do this, but I'm not sure if it's more appropriate, this solution  looks fine to me.